### ISSUE TITLE
Bad error message: "Patterns are incomplete" when matching on polymorphic value

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#238]
We currently get a `Patterns are incomplete` error when pattern matching on a polymorphic value. This is confusing, since the real problem has nothing to do with the completeness of patterns. Here is a small example illustrating this:

``` fsharp
type box = | Box : int -> box

val silly : a:Type -> a -> int
let silly (a:Type) x = let Box y = x in y+1
```

Error message:

```
bug238.fst(6,40-6,43): Patterns are incomplete
```

The range points at the body of the let (`y+1`), again, not where the problem is.

**Comment by msprotz** [comment]
What's the workaround in that case then?

**Comment by catalin-hritcu** [comment]
Not matching on a value that have a type variable as its type?

**Comment by vkanne-msft** [comment]
Cannot reproduce the issue. Closing for now.

**Comment by nikswamy** [comment]
@catalin-hritcu's original program actually seems to verify with F* master, which looks like a pretty bad bug to me.

It rightfully fails to verify in my nik_pat branch with a 

```
bug238.fst(6,27-6,32): (Error 114) Type of pattern (Bug238.box) does not match type of scrutinee (a)
```

**Comment by nikswamy** [comment]
for the record, this is easily turned into a proof of false:

```
module Bug238

type box = | Box : False -> box

val silly : a:Type -> a -> False
let silly (a:Type) x = let Box y = x in y
```

**Comment by nikswamy** [comment]
Fixed since 5176a7b35d38b81c864307b602c8034c0a8c3785
